src/OpenLoco/IndustryManager.cpp

```diff
@@ -190,18 +190,238 @@ namespace OpenLoco::IndustryManager
+            auto* surface = tile.surface();
+            auto height = surface->baseHeight();
+            lowest = std::min(lowest, height);
+            if (surface->slope())
+            {
+                height += 16;
+                if (surface->isSlopeDoubleHeight())
+                {
+                    height += 16;
+                }
+            }
+            highest = std::max(highest, height);
+        }
+        return highest - lowest;
+    }
+
     // 0x004599B3
     static std::optional<Map::Pos2> findRandomNewIndustryLocation(const uint8_t indObjId)
     {
-        registers regs;
-        regs.edx = indObjId;
-        call(0x004599B3, regs);
-        if (regs.ax == -1)
+        auto* indObj = ObjectManager::get<IndustryObject>(indObjId);
+        for (auto i = 0; i < 250; ++i)
         {
-            return std::nullopt;
-        }
+            // Replace the below with this after validating the function
+            // Map::Pos2 randomPos{
+            //     Map::TilePos2(gPrng().randNext(Map::kMapRows), gPrng().randNext(Map::kMapColumns))
+            // };
+            const auto randomNum = gPrng().randNext();
+
+            Map::Pos2 randomPos{
+                Map::TilePos2(
+                    (((randomNum >> 16) * Map::kMapRows) >> 16),
+                    (((randomNum & 0xFFFF) * Map::kMapColumns) >> 16))
+            };
+
+            if (isTooCloseToNearbyIndustries(randomPos))
+            {
+                continue;
+            }
+
+            if (indObj->flags & IndustryObjectFlags::builtInClusters)
+            {
+                if (isOutwithCluster(randomPos, indObjId))
+                {
+                    continue;
+                }
+            }
 
-        return Map::Pos2{ regs.ax, regs.cx };
+            if (indObj->flags & IndustryObjectFlags::builtOnHighGround)
+            {
+                auto tile = Map::TileManager::get(randomPos);
+                auto* surface = tile.surface();
+                if (surface == nullptr || surface->baseZ() < 48)
+                {
+                    continue;
+                }
+            }
+            if (indObj->flags & IndustryObjectFlags::builtOnLowGround)
+            {
+                auto tile = Map::TileManager::get(randomPos);
+                auto* surface = tile.surface();
+                if (surface == nullptr || surface->baseZ() > 56)
+                {
+                    continue;
+                }
+            }
+            if (indObj->flags & IndustryObjectFlags::builtOnSnow)
+            {
+                auto tile = Map::TileManager::get(randomPos);
+                auto* surface = tile.surface();
+                auto* climateObj = ObjectManager::get<ClimateObject>();
+                if (surface == nullptr || surface->baseZ() < climateObj->summerSnowLine)
+                {
+                    continue;
+                }
+            }
+            if (indObj->flags & IndustryObjectFlags::builtBelowSnowLine)
```

Comment:
a log of these `if` statements are really similar and could probably be combined into a function

Reply:
A future refactor could map flags to functions, perhaps? Not necessary for this PR, imo.

Reply:
I'm not sure you could really combine them that much without wasting time computing things that aren't used.